datafusion/optimizer/src/push_down_filter.rs

```diff
@@ -500,302 +387,336 @@ fn optimize_join(
+                        replace_map.insert(
+                            union.schema.fields().get(i).unwrap().qualified_name(),
+                            Expr::Column(field.qualified_column()),
+                        );
+                    }
 
-            optimize_join(state, plan, left, right, on_filters)
-        }
-        LogicalPlan::TableScan(TableScan {
-            source,
-            projected_schema,
-            filters,
-            projection,
-            table_name,
-            fetch,
-        }) => {
-            let mut used_columns = HashSet::new();
-            let mut new_filters = filters.clone();
-
-            for (filter_expr, cols) in &state.filters {
-                let (preserve_filter_node, add_to_provider) =
-                    match source.supports_filter_pushdown(filter_expr)? {
-                        TableProviderFilterPushDown::Unsupported => (true, false),
-                        TableProviderFilterPushDown::Inexact => (true, true),
-                        TableProviderFilterPushDown::Exact => (false, true),
-                    };
-
-                if preserve_filter_node {
-                    used_columns.extend(cols.clone());
+                    let push_predicate =
+                        replace_cols_by_name(filter.predicate().clone(), &replace_map)?;
+                    inputs.push(Arc::new(LogicalPlan::Filter(Filter::try_new(
+                        push_predicate,
+                        input.clone(),
+                    )?)))
                 }
-
-                if add_to_provider {
-                    // Don't add expression again if it's already present in
-                    // pushed down filters.
-                    if new_filters.contains(filter_expr) {
-                        continue;
+                LogicalPlan::Union(Union {
+                    inputs,
+                    schema: plan.schema().clone(),
+                })
+            }
+            LogicalPlan::Aggregate(agg) => {
+                // An aggregate's aggregate columns are _not_ filter-commutable => collect these:
+                // * columns whose aggregation expression depends on
+                // * the aggregation columns themselves
+
+                // construct set of columns that `aggr_expr` depends on
+                let mut used_columns = HashSet::new();
+                exprlist_to_columns(&agg.aggr_expr, &mut used_columns)?;
+                let agg_columns = agg
+                    .aggr_expr
+                    .iter()
+                    .map(|x| Ok(Column::from_name(x.display_name()?)))
+                    .collect::<Result<HashSet<_>>>()?;
+                used_columns.extend(agg_columns);
+
+                let predicates = utils::split_conjunction_owned(utils::cnf_rewrite(
+                    filter.predicate().clone(),
+                ));
+
+                let mut keep_predicates = vec![];
+                let mut push_predicates = vec![];
```

Comment:
You can use the `partition()` method from `std::iter::Iterator ` which can split a iter/vec into two parts based on a true/false predicate, this is more FP style.

https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition

Reply:
🥲 I don't know how to use `partition()` when meet result in closure